composer/datasets/streaming/dataset.py

```diff
@@ -392,6 +327,22 @@ def __iter__(self) -> Iterator[Any]:
             Iterator[Any]: Each sample.
         """
         Thread(target=self.download, daemon=True).start()
-
-        for idx in self._iter_ids():
-            yield self[idx]
+        num_workers = dist.get_local_world_size()
+        rank = dist.get_local_rank()
+        sbs = int(self._shuffle_buffer_size)
+        while self._batch_count < self.index.total_samples:
+            if self._batch_count < self._restored_batch_count:
+                self._batch_count += 1
+                yield None
+            try:
+                idx = self.shuffler.shuffle_sample(self._batch_count, num_workers, rank, sbs) \
+                    if self.shuffle else self._batch_count
+                yield self[idx]
+                self._batch_count += 1
```

Comment:
I think there's a bit of a naming and/or logic issue here, the DataLoader requests _samples_ from individual dataset workers, not _batches_. 

E.g. If I have a dataloader with device_batch_size=8, and I have 3 workers, then I am going to request samples round robin from the 3 workers (with independent dataset copies) and get samples like:
```
[w0_s0, w1_s0, w2_s0, w0_s1, w1_s1, w2_s1, w0_s2]
```

Reply:
ah, meant to write sample_count. apologies